File: Editorial/0000-0100/0084_Largest_Rectangle_in_Histogram/C/3_Divide_and_Conquer_Approach.c
int max(int a, int b) { return (a > b) ? a : b; }
int calculateArea(int* heights, int start, int end) {
    if (start > end) return 0;
    int min_index = start;
    for (int i = start; i <= end; i++)
        if (heights[min_index] > heights[i]) min_index = i;
    return max(heights[min_index] * (end - start + 1),
               max(calculateArea(heights, start, min_index - 1),
                   calculateArea(heights, min_index + 1, end)));
}
int largestRectangleArea(int* heights, int heightsSize) {
    return calculateArea(heights, 0, heightsSize - 1);
}

File: Editorial/0000-0100/0084_Largest_Rectangle_in_Histogram/CSharp/3_Divide_and_Conquer_Approach.cs
public class Solution {
    public int CalculateArea(int[] heights, int start, int end) {
        if (start > end)
            return 0;
        int min_index = start;
        for (int i = start; i <= end; i++)
            if (heights[min_index] > heights[i])
                min_index = i;
        return Math.Max(heights[min_index] * (end - start + 1),
                        Math.Max(CalculateArea(heights, start, min_index - 1),
                                 CalculateArea(heights, min_index + 1, end)));
    }

    public int LargestRectangleArea(int[] heights) {
        return CalculateArea(heights, 0, heights.Length - 1);
    }
}
